### PR TITLE
refactor: simplified wallet setup

### LINKED ISSUE #51: Simplified Wallet Setup with the New API
# New wallet initialization API

The current implementation of the `Wallet` class lacks readability and user-friendliness, especially during instance creation. Developers are currently forced to use the constructor directly for initialization, with options like `Wallet.fromMnemonic`, `Wallet.fromKeystore`, or `Wallet.createRandom` instances method available separately to initialize `Wallet`. This approach leads to confusion and makes the code less maintainable. A more streamlined and intuitive design is needed to enhance the readability and developer experience when setting up a `Wallet` instance.

## Solution:

The SDK should expose unified static methods responsible for both creating and initializing a `Wallet` instance. These methods should handle various initialization options, such as mnemonic, keystore, or random generation. Meanwhile, the `constructor` should be reserved for creating a `Wallet` instance using the private key and curve. This approach aims to streamline the initialization process, making it more intuitive and developer-friendly.

### PR DESCRIPTION
**Pull request by sarvalabs-harshrastogi**
# Simplified Wallet Setup and Initialization

This pull request introduces significant enhancements to the `js-moi-wallet` module, simplifying the wallet setup and initialization APIs for a more user-friendly experience.

## Changes

The following key changes have been implemented:

- **`constructor`**: Now, a Wallet instance is created using the private key, serving as both the creation and initialization step.

- **`Wallet.fromMnemonic`**: Enables the creation of a wallet using a mnemonic and a specified path.

- **`Wallet.fromMnemonicSync`**: Synchronously creates a wallet using a mnemonic and a path.

- **`Wallet.fromKeystore`**: Allows the creation of a wallet instance using a provided keystore and password.

- **`Wallet.createRandom`**: Introduces a method to create a wallet instance using a randomly generated mnemonic.

- **`Wallet.createRandomSync`**: Provides a synchronous alternative for creating a wallet instance with a randomly generated mnemonic.

These changes collectively simplify the wallet setup process and contribute to a more intuitive and streamlined API structure.

- closes #51 

## Changes include
- [ ] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [x] Breaking change (change that is not backwards-compatible and/or changes current functionality)

## Checklist
- [X] I have assigned this PR to myself
- [X] I have added at least 1 reviewer
- [X] I have tested this code
- [X] I have updated the README and other relevant documents (guides...)

### CODE REVIEW

File: docs/source/logic.rst [hunk 1]
```diff
@@ -128,9 +128,10 @@ applications on the MOI network.
     // Example
     const initWallet = async () => {
         const mnemonic = "mother clarify push liquid ordinary social track ...";
-        const provider = new JsonRpcProvider("http://localhost:1600/");
-        const wallet = new Wallet(provider);
-        await wallet.fromMnemonic(mnemonic);
+        const wallet = await Wallet.fromMnemonic(mnemonic);
+
+        wallet.connect(new JsonRpcProvider("http://localhost:1600/"));
+
```
Comment (sarvalabs-gokul) [inline docs/source/logic.rst:134]: nitpick, please create a variable for provider and use it to enhance the readability.

File: packages/js-moi-logic/__tests__/logic.test.ts [hunk 2]
```diff
@@ -17,8 +17,8 @@ const PROVIDER = new JsonRpcProvider(HOST);
 let wallet: Wallet;
 
 beforeAll(async () => {
-    wallet = new Wallet(PROVIDER);
-    await wallet.fromMnemonic(MNEMONIC, DEVIATION_PATH);
+    wallet = await Wallet.fromMnemonic(MNEMONIC, DEVIATION_PATH);
```
Comment (sarvalabs-gokul) [inline packages/js-moi-logic/__tests__/logic.test.ts:20]: what is DEVIATION_PATH.

File: packages/js-moi-wallet/README.md [hunk 3]
```diff
@@ -34,10 +34,14 @@ npm install js-moi-wallet
...
-        const path = "m/44'/7567'/0'/0/1";
-        await wallet.fromMnemonic(mnemonic, path);
+         const initWallet = async () => {
+        const mnemonic = "mother clarify push liquid ordinary social track ...";
+        const wallet = await Wallet.fromMnemonic(mnemonic);
+
+        wallet.connect(new JsonRpcProvider("http://localhost:1600/"));
+
+        return wallet;
+    }
```
Comment (sarvalabs-gokul) [inline packages/js-moi-wallet/README.md:44]: nitpick, fix the indentation issue.

File: packages/js-moi-wallet/dist/wallet.d.ts [hunk 4]
```diff
@@ -1,40 +1,43 @@
...
-     * Initializes the wallet with a private key, mnemonic, and curve.
-     *
-     * @param {Buffer} key - The private key as a Buffer.
-     * @param {string} curve - The elliptic curve algorithm used for key generation.
-     * @param {string} mnemonic - The mnemonic associated with the wallet. (optional)
-     * @throws {Error} if the key is undefined or if an error occurs during the
-     * initialization process.
-     */
-    load(key: Buffer, curve: string, mnemonic?: string): void;
+    constructor(key: Buffer, curve: string, mnemonic?: string);
```
Comment (sarvalabs-gokul) [inline packages/js-moi-wallet/dist/wallet.d.ts:34]: do not mix both private key and mnemonic. As there is already a static method to create a wallet from a mnemonic, I don't see any need for a mnemonic in the constructor.

File: packages/js-moi-wallet/dist/wallet.d.ts [hunk 5]
```diff
@@ -45,58 +48,44 @@ export declare class Wallet extends Signer {
...
      *
      * @returns {string} The address as a string.
      */
     getAddress(): string;
+    /**
+     * Address associated with the wallet.
+     *
+     * @readonly
+     */
+    get address(): string;
```
Comment (sarvalabs-gokul) [inline packages/js-moi-wallet/dist/wallet.d.ts:88]: why do we need two methods

**Review by sarvalabs-gokul — CHANGES_REQUESTED** [review]

File: __tests__/e2e/interactions.test.ts [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            } catch (error) {
+                expect(error).not.toBeNull();
+
+                if (error instanceof CustomError) {
+                    expect(error.params.receipt).toBeDefined();
+                }
+            }
+        });
+    });
+});
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply __tests__/e2e/interactions.test.ts:372]: a36b11b60f5391536a65926f9450046806d84589

File: docs/source/hierarchical-deterministic-wallet.rst [hunk 7]
```diff
@@ -44,190 +44,93 @@ The ``Keystore`` interface represents a keystore object. It has the following pr
...
-            "kdf": "scrypt",
-            "kdfparams": {
-                ...
-            },
-            "mac": "..."
-        }
-    */
-
-.. autofunction:: Wallet#fromMnemonic
+        const wallet = new Wallet(Buffer.from(privateKey, "hex"), CURVE.SECP256K1);
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/hierarchical-deterministic-wallet.rst:67]: refactored in: 8d175a81c59d9b615201a41ac64f4adfee3bec44 a36b11b60f5391536a65926f9450046806d84589

view file: https://github.com/sarvalabs/js-moi-sdk/blob/8d175a81c59d9b615201a41ac64f4adfee3bec44/packages/js-moi-wallet/src/wallet.ts#L90-L123

File: docs/source/hierarchical-deterministic-wallet.rst [hunk 8]
```diff
@@ -44,190 +44,93 @@ The ``Keystore`` interface represents a keystore object. It has the following pr
...
-        },
-        "mac": "..."
-    }
-    wallet.fromKeystore(keystore, "CZ%90$DI");
-
-.. autofunction:: Wallet#mnemonic
+    .. code-block:: javascript
 
-.. code-block:: javascript
+        const keystore = { ... };
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/hierarchical-deterministic-wallet.rst]: a36b11b60f5391536a65926f9450046806d84589

File: docs/source/hierarchical-deterministic-wallet.rst [hunk 9]
```diff
@@ -44,190 +44,93 @@ The ``Keystore`` interface represents a keystore object. It has the following pr
...
-    console.log(address);
-
-    >> 0x87925...
-
-.. autofunction:: Wallet#connect
+  const message = "Hello, MOI";
+  const algo = wallet.signingAlgorithms["ecdsa_secp256k1"];
 
-.. code-block:: javascript
+  const signature = wallet.sign(Buffer.from(message), algo);
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/hierarchical-deterministic-wallet.rst:159]: a36b11b60f5391536a65926f9450046806d84589

File: docs/source/logic.rst (hunk 1 above)
Comment (sarvalabs-harshrastogi) [reply docs/source/logic.rst:134]: a36b11b60f5391536a65926f9450046806d84589

File: packages/js-moi-logic/__tests__/logic.test.ts (hunk 2 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/__tests__/logic.test.ts]: a36b11b60f5391536a65926f9450046806d84589

File: packages/js-moi-wallet/README.md (hunk 3 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-wallet/README.md]: a36b11b60f5391536a65926f9450046806d84589

File: packages/js-moi-wallet/dist/wallet.d.ts (hunk 4 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-wallet/dist/wallet.d.ts]: refactored in: a36b11b60f5391536a65926f9450046806d84589

--- 
- Remove storing of mnemonic in `__vault` in constructor function:

  https://github.com/sarvalabs/js-moi-sdk/blob/8d175a81c59d9b615201a41ac64f4adfee3bec44/packages/js-moi-wallet/src/wallet.ts#L90-L123

File: packages/js-moi-wallet/dist/wallet.d.ts (hunk 5 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-wallet/dist/wallet.d.ts:88]: The `Wallet` class extends the `Signer` abstract class, thereby providing a concrete implementation of the `getAddress` method required by the abstract class. Additionally, the `Wallet` class includes an `address` getter method, which returns the same value as the `getAddress` method, serving as a convenient way to access the associated wallet address.

**Review by sarvalabs-gokul — APPROVED** [review]

**Pull request merged**